File: 41. First Missing Positive.py
class Solution(object):
    def firstMissingPositive(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if len(nums)==0:
            return 1
        for i in range(len(nums)):
            nums[i]=nums[i]-1

        for i in range(len(nums)):
            while nums[i]!=i and nums[i]>=0 and nums[i]<len(nums) and nums[i]!=nums[nums[i]]:
                temp=nums[i]
                nums[i]=nums[temp]
                nums[temp]=temp;

        for i in range(len(nums)):
            if nums[i]!=i:
                return i+1

        return len(nums)+1

File: 41. First Missing Positive.cpp
class Solution {
public:
    int firstMissingPositive(vector<int>& nums) {
        if(nums.size()==0) return 1;
        for(int i=0;i<nums.size();i++) nums[i]=nums[i]-1;
        for(int i=0;i<nums.size();i++){
          while(nums[i]>=0 && nums[i]<nums.size() && nums[i]!=i && nums[i]!=nums[nums[i]]){
            int temp=nums[i];
            nums[i]=nums[temp];
            nums[temp]=temp;
          }
        }

        for(int i=0;i<nums.size();i++)
          if(nums[i]!=i) return i+1;

        return nums.size()+1;
    }
};
